### ISSUE TITLE
Build failure: `pkgsStatic.perl` (Perl interpreter statically linked to Musl libc)

### ISSUE THREAD

**Issue by YoshiRulz** [NixOS/nixpkgs#295608]
### Steps To Reproduce

1. e.g. `nix build -I nixpkgs=flake:github:NixOS/nixpkgs/23.11 nixpkgs#pkgsStatic.perl538`

It still affects `master` as of 2024-03-12. edit: Changed title for clarity; `pkgsMusl.perl` is fine.

### Build log

<details>
<summary>tail of build log</summary>

```
make -C dist/threads-shared/ PERL_CORE=1 LIBPERL=libperl.a LINKTYPE=static static
make[2]: Entering directory '/build/perl-5.38.2/dist/threads-shared'
"../../miniperl_top" "-I../../lib" "../../lib/ExtUtils/xsubpp"  -typemap '/build/perl-5.38.2/dist/threads-shared/../../lib/ExtUtils/typemap'  shared.xs > shared.xsc
mv shared.xsc shared.c
x86_64-unknown-linux-musl-gcc -c   -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -Wall -Wno-use-after-free -fwrapv -fno-strict-aliasing   -DVERSION=\"1.68\" -DXS_VERSION=\"1.68\" -fPIC -Wno-unused-function "-I../.."   shared.c
In file included from /nix/store/5m5zr8z4m8wx4v2c6w3fwcg07igrwzjy-libxcrypt-static-x86_64-unknown-linux-musl-4.4.36/include/crypt.h:22,
                 from ../../reentr.h:124,
                 from ../../op.h:700,
                 from ../../perl.h:4542,
                 from shared.xs:131:
/nix/store/0nfn0979vnb1w48rzm5mkx9zvbrkmxpp-x86_64-unknown-linux-musl-gcc-13.2.0/x86_64-unknown-linux-musl/sys-include/sys/cdefs.h:1:2: warning: #warning usage of non-standard #include <sys/cdefs.h> is deprecated [-Wcpp]
    1 | #warning usage of non-standard #include <sys/cdefs.h> is deprecated
      |  ^~~~~~~
In file included from ../../perl.h:7816:
../../sv_inline.h:566:1: warning: 'always_inline' function might not be inlinable [-Wattributes]
  566 | Perl_SvPVXtrue(pTHX_ SV *sv)
      | ^~~~~~~~~~~~~~
../../sv_inline.h:536:1: warning: 'always_inline' function might not be inlinable [-Wattributes]
  536 | Perl_newSV_type_mortal(pTHX_ const svtype type)
      | ^~~~~~~~~~~~~~~~~~~~~~
rm -f "../../lib/auto/threads/shared/shared.a"
x86_64-unknown-linux-musl-ar cr "../../lib/auto/threads/shared/shared.a" shared.o
x86_64-unknown-linux-musl-ranlib "../../lib/auto/threads/shared/shared.a"
chmod 755 ../../lib/auto/threads/shared/shared.a
make[2]: Leaving directory '/build/perl-5.38.2/dist/threads-shared'
./miniperl_top statars B Devel::Peek Fcntl File::DosGlob File::Glob Hash::Util Hash::Util::FieldHash I18N::Langinfo Opcode POSIX PerlIO::encoding PerlIO::mmap PerlIO::scalar PerlIO::via SDBM_File Sys::Hostname XS::APItest XS::Typemap attributes mro re Compress::Raw::Bzip2 Compress::Raw::Zlib Digest::MD5 Digest::SHA Encode Filter::Util::Call IPC::SysV MIME::Base64 Math::BigInt::FastCalc List::Util Socket Sys::Syslog Time::Piece Unicode::Collate Data::Dumper Devel::PPPort IO Cwd Storable Time::HiRes Unicode::Normalize threads threads::shared > static.list
./miniperl_top extlibs B Devel::Peek Fcntl File::DosGlob File::Glob Hash::Util Hash::Util::FieldHash I18N::Langinfo Opcode POSIX PerlIO::encoding PerlIO::mmap PerlIO::scalar PerlIO::via SDBM_File Sys::Hostname XS::APItest XS::Typemap attributes mro re Compress::Raw::Bzip2 Compress::Raw::Zlib Digest::MD5 Digest::SHA Encode Filter::Util::Call IPC::SysV MIME::Base64 Math::BigInt::FastCalc List::Util Socket Sys::Syslog Time::Piece Unicode::Collate Data::Dumper Devel::PPPort IO Cwd Storable Time::HiRes Unicode::Normalize threads threads::shared > ext.libs
x86_64-unknown-linux-musl-gcc -Wl,-E -o perl perlmain.o libperl.a lib/auto/B/B.a lib/auto/Devel/Peek/Peek.a lib/auto/Fcntl/Fcntl.a lib/auto/File/DosGlob/DosGlob.a lib/auto/File/Glob/Glob.a lib/auto/Hash/Util/Util.a lib/auto/Hash/Util/FieldHash/FieldHash.a lib/auto/I18N/Langinfo/Langinfo.a lib/auto/Opcode/Opcode.a lib/auto/POSIX/POSIX.a lib/auto/PerlIO/encoding/encoding.a lib/auto/PerlIO/mmap/mmap.a lib/auto/PerlIO/scalar/scalar.a lib/auto/PerlIO/via/via.a lib/auto/SDBM_File/SDBM_File.a lib/auto/Sys/Hostname/Hostname.a lib/auto/XS/APItest/APItest.a lib/auto/XS/Typemap/Typemap.a lib/auto/attributes/attributes.a lib/auto/mro/mro.a lib/auto/re/re.a lib/auto/Compress/Raw/Bzip2/Bzip2.a lib/auto/Compress/Raw/Zlib/Zlib.a lib/auto/Digest/MD5/MD5.a lib/auto/Digest/SHA/SHA.a lib/auto/Encode/Encode.a lib/auto/Filter/Util/Call/Call.a lib/auto/IPC/SysV/SysV.a lib/auto/MIME/Base64/Base64.a lib/auto/Math/BigInt/FastCalc/FastCalc.a lib/auto/List/Util/Util.a lib/auto/Socket/Socket.a lib/auto/Sys/Syslog/Syslog.a lib/auto/Time/Piece/Piece.a lib/auto/Unicode/Collate/Collate.a lib/auto/Data/Dumper/Dumper.a lib/auto/Devel/PPPort/PPPort.a lib/auto/IO/IO.a lib/auto/Cwd/Cwd.a lib/auto/Storable/Storable.a lib/auto/Time/HiRes/HiRes.a lib/auto/Unicode/Normalize/Normalize.a lib/auto/threads/threads.a lib/auto/threads/shared/shared.a -lm -lcrypt -lpthread -ldl -lm -L/nix/store/ljqvlm46wa0v3r39kii8k2gzwk6jxybs-zlib-static-x86_64-unknown-linux-musl-1.3.1/lib -lz -lrt -lpthread
/nix/store/sqii1hjj7rr1d5c5wwxlswqrv39nispr-x86_64-unknown-linux-musl-binutils-2.41/bin/x86_64-unknown-linux-musl-ld: lib/auto/re/re.a(re_comp.o): in function `Perl_reg_add_data':
re_comp.c:(.text+0xa390): multiple definition of `Perl_reg_add_data'; libperl.a(regcomp.o):regcomp.c:(.text+0x70a0): first defined here
/nix/store/sqii1hjj7rr1d5c5wwxlswqrv39nispr-x86_64-unknown-linux-musl-binutils-2.41/bin/x86_64-unknown-linux-musl-ld: lib/auto/re/re.a(re_comp.o): in function `Perl_populate_anyof_bitmap_from_invlist':
re_comp.c:(.text+0xa8e0): multiple definition of `Perl_populate_anyof_bitmap_from_invlist'; libperl.a(regcomp.o):regcomp.c:(.text+0x7b60): first defined here
/nix/store/sqii1hjj7rr1d5c5wwxlswqrv39nispr-x86_64-unknown-linux-musl-binutils-2.41/bin/x86_64-unknown-linux-musl-ld: lib/auto/re/re.a(re_comp.o): in function `Perl_add_above_Latin1_folds':
re_comp.c:(.text+0xab10): multiple definition of `Perl_add_above_Latin1_folds'; libperl.a(regcomp.o):regcomp.c:(.text+0x7ce0): first defined here
/nix/store/sqii1hjj7rr1d5c5wwxlswqrv39nispr-x86_64-unknown-linux-musl-binutils-2.41/bin/x86_64-unknown-linux-musl-ld: lib/auto/re/re.a(re_comp.o): in function `Perl_set_ANYOF_arg':
re_comp.c:(.text+0xb020): multiple definition of `Perl_set_ANYOF_arg'; libperl.a(regcomp.o):regcomp.c:(.text+0x8160): first defined here
/nix/store/sqii1hjj7rr1d5c5wwxlswqrv39nispr-x86_64-unknown-linux-musl-binutils-2.41/bin/x86_64-unknown-linux-musl-ld: lib/auto/re/re.a(re_comp.o): in function `Perl_get_ANYOFM_contents':
re_comp.c:(.text+0xcf50): multiple definition of `Perl_get_ANYOFM_contents'; libperl.a(regcomp.o):regcomp.c:(.text+0x9d70): first defined here
/nix/store/sqii1hjj7rr1d5c5wwxlswqrv39nispr-x86_64-unknown-linux-musl-binutils-2.41/bin/x86_64-unknown-linux-musl-ld: lib/auto/re/re.a(re_comp.o): in function `Perl_get_ANYOFHbbm_contents':
re_comp.c:(.text+0xd020): multiple definition of `Perl_get_ANYOFHbbm_contents'; libperl.a(regcomp.o):regcomp.c:(.text+0x9e40): first defined here
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:154: perl] Error 1
make[1]: Leaving directory '/build/perl-5.38.2'
make: *** [Makefile:87: all] Error 2
```
</details>

[full log](https://gist.github.com/YoshiRulz/9386c76c8d2a59deb74a0edd114f1327)

### Additional context

I'm aiming for `pkgsStatic.wine64Packages.minimal`, which may be a fool's errand but it's worth a try.

### Notify maintainers

@stigtsp @zakame @dasJ (and Eelco?)

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.38, NixOS, 23.05 (Stoat), 23.05.1875.8df7a67abaf`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by bltavares** [comment]
There is an [upstream issue](https://github.com/Perl/perl5/issues/21550), with a [patch](https://github.com/Perl/perl5/commit/ba6e2c38aafc23cf114f3ba0d0ff3baead34328b) reported to fix it.

The commit has been already applied to Perl 5.39.2, so I'm unsure if it's better to attempt to patch 5.38 (current version), or to attempt to upgrade nixpkgs to 5.39.9 instead.

**Comment by stigtsp** [comment]
> The commit has been already applied to Perl 5.39.2, so I'm unsure if it's better to attempt to patch 5.38 (current version), or to attempt to upgrade nixpkgs to 5.39.9 instead.

Nice find! I think we should conditionally patch this in nixpkgs, as 5.38 will stay around for a while.

**Comment by bltavares** [comment]
I'm not very familiar with nix nor nixpkgs contribution process, but I'm trying it locally at this moment (and the world is being recompiled, so it's some sign of success, I guess).

If I'm able to have a successful build I'll submit a PR soon.

**Comment by stigtsp** [comment]
I'm happy to help out of course. Do you have a branch somewhere with those changes?

**Comment by bltavares** [comment]
@stigtsp thanks for offering to help.

I've applied [the patch regardless](https://github.com/NixOS/nixpkgs/compare/master...bltavares:nixpkgs-upstream:pkgsStatic.perl) of it being a `stdenv.hostPlatform.isStatic` as upstream has changed for all stdlibs, but it can be checked if it would be better.

**Comment by stigtsp** [comment]
If you'd want to set up a PR I think that would be best, also we should be able to avoid mass rebuilds as mentioned here: https://github.com/NixOS/nixpkgs/commit/1cbba6d9bd08fe8c5947a1d4c2a6adcdae7a4cd4#r140306610

**Comment by bltavares** [comment]
I'm still waiting on my Linux VM to finish compiling the world, which sounds like a good scenario. 

Meanwhile I came across a chain of ~yaks~ issues on MacOS that prevents `pkgsStatic.perl` to cross-compile due to case-sensitiveness of the filesystem, even after enabling it. https://github.com/NixOS/nixpkgs/issues/113985#issuecomment-785024419 -> https://github.com/NixOS/nix/issues/1446 -> https://github.com/NixOS/nix/issues/9318


PS: Another ~yak~ issue appeared after being able to move `nix-daemon` to a `TMPDIR` that is case-sensitive:

```
       > Guessing build platform ... aarch64-apple-darwin
       > Configuring for a native build
       > Checking perl version ... perl5-5.38.2
       > Checking for cc ... aarch64-apple-darwin-clang
       > Checking for ar ... aarch64-apple-darwin-ar
       > Checking for nm ... aarch64-apple-darwin-nm
       > Checking for ranlib ... aarch64-apple-darwin-ranlib
       > Checking for readelf ... none found
       > ERROR: Cannot find readelf
```

I'm waiting for the Linux build to conclude in order to open a Pull Request.

**Comment by YoshiRulz** [comment]
I believe this is presently fixed in `staging` but not yet `master`? Thanks to those who contributed to that.

**Comment by stigtsp** [comment]
This seems fixed in staging (tested on 888bf6922bdb7faee5efffec3d1beffeaf64ef2e):

```
$ nom build .#pkgsStatic.perl
...
$ result/bin/perl -MConfig -E 'say $Config{archname}'
x86_64-linux-thread-multi
...
$ file result/bin/perl
result/bin/perl: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, not stripped
```
